src/cascadia/WindowsTerminal/NonClientIslandWindow.cpp

```diff
@@ -461,342 +428,120 @@ RECT NonClientIslandWindow::GetMaxWindowRectInPixels(const RECT* const prcSugges
-
-            // Draw the bottom window border
-            clientRect = { 0, cy - yPos, cx, cy };
-            ::FillRect(hdc.get(), &clientRect, _backgroundBrush.get());
-
-            // Draw the right window border
-            clientRect = { cx - xPos, 0, cx, cy };
-            ::FillRect(hdc.get(), &clientRect, _backgroundBrush.get());
-        }
-
         return 0;
     }
 
-    case WM_NCLBUTTONDOWN:
-    case WM_NCLBUTTONUP:
-    case WM_NCMBUTTONDOWN:
-    case WM_NCMBUTTONUP:
-    case WM_NCRBUTTONDOWN:
-    case WM_NCRBUTTONUP:
-    case WM_NCXBUTTONDOWN:
-    case WM_NCXBUTTONUP:
+    const auto topBorderHeight = GetTopBorderHeight();
+
+    if (ps.rcPaint.top < topBorderHeight)
     {
-        // If we clicked in the titlebar, raise an event so the app host can
-        // dispatch an appropriate event.
-        _DragRegionClickedHandlers();
-        break;
+        RECT rcTopBorder = ps.rcPaint;
+        rcTopBorder.bottom = topBorderHeight;
+
+        // To show the original top border, we have to paint on top of it with
+        // the alpha component set to 0. This page recommends to paint the area
+        // in black using the stock BLACK_BRUSH to do this:
+        // https://docs.microsoft.com/en-us/windows/win32/dwm/customframe#extending-the-client-frame
+        ::FillRect(hdc.get(), &rcTopBorder, GetStockBrush(BLACK_BRUSH));
     }
 
-    case WM_WINDOWPOSCHANGING:
+    if (ps.rcPaint.bottom > topBorderHeight)
     {
-        // Enforce maximum size here instead of WM_GETMINMAXINFO. If we return
-        // it in WM_GETMINMAXINFO, then it will be enforced when snapping across
-        // DPI boundaries (bad.)
-        LPWINDOWPOS lpwpos = reinterpret_cast<LPWINDOWPOS>(lParam);
-        if (lpwpos == nullptr)
-        {
-            break;
-        }
-        if (_HandleWindowPosChanging(lpwpos))
+        RECT rcRest = ps.rcPaint;
+        rcRest.top = topBorderHeight;
+
+        const auto backgroundBrush = _titlebar.Background();
+        const auto backgroundSolidBrush = backgroundBrush.as<Media::SolidColorBrush>();
+        const auto backgroundColor = backgroundSolidBrush.Color();
+        const auto color = RGB(backgroundColor.R, backgroundColor.G, backgroundColor.B);
+
+        if (!_backgroundBrush || color != _backgroundBrushColor)
         {
-            return 0;
+            // Create brush for titlebar color.
+            _backgroundBrush = wil::unique_hbrush(CreateSolidBrush(color));
         }
-        else
+
+        // To hide the original title bar, we have to paint on top of it with
+        // the alpha component set to 255. This is a hack to do it with GDI.
+        // See NonClientIslandWindow::_UpdateFrameMargins for more information.
+        HDC opaqueDc;
+        BP_PAINTPARAMS params = { sizeof(params), BPPF_NOCLIP | BPPF_ERASE };
+        HPAINTBUFFER buf = BeginBufferedPaint(hdc.get(), &rcRest, BPBF_TOPDOWNDIB, &params, &opaqueDc);

```

Comment:
I don't really know how GDI works so I found the first solution that worked on Google to do this but maybe there is a better solution.